pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,281 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"reflect"
+	"sort"
+	"strings"
+	"testing"
+
+	cmg "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+const (
+	httpPattern            = "/metrics"
+	addr                   = "localhost:0"
+	processStartTimeMetric = "process_start_time_seconds"
+)
+
+func initMgr() (MetricsManager, *http.Server) {
+	mgr := NewMetricsManager()
+	mux := http.NewServeMux()
+	err := mgr.PrepareMetricsPath(mux, httpPattern, nil)
+	if err != nil {
+		log.Fatalf("failed to start serving [%v]", err)
+	}
+	l, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Fatalf("failed to listen on address[%s], error[%v]", addr, err)
+	}
+	srv := &http.Server{Addr: l.Addr().String(), Handler: mux}
+	go func() {
+		if err := srv.Serve(l); err != http.ErrServerClosed {
+			log.Fatalf("failed to start endpoint at:%s/%s, error: %v", addr, httpPattern, err)
+		}
+	}()
+
+	return mgr, srv
+}
+
+func shutdown(srv *http.Server) {
+	if err := srv.Shutdown(context.Background()); err != nil {
+		panic(err)
+	}
+}
+
+func TestNew(t *testing.T) {
+	mgr, srv := initMgr()
+	defer shutdown(srv)
+	if mgr == nil {
+		t.Errorf("failed testing new")
+	}
+}
+
+func TestIncrementCount(t *testing.T) {
+	mgr, srv := initMgr()
+	srvAddr := "http://" + srv.Addr + httpPattern
+	defer shutdown(srv)
+	mgr.IncrementCount("test_result_1")
+	mgr.IncrementCount("test_result_2")
+	mgr.IncrementCount("test_result_2")
+
+	expected :=
+		`# HELP process_start_time_seconds [ALPHA] Start time of the process since unix epoch in seconds.
+        # TYPE process_start_time_seconds gauge
+        process_start_time_seconds 0
+# HELP volume_data_source_validator_volume_data_source_validator_operation_count [ALPHA] Number of validations operations by result
+# TYPE volume_data_source_validator_volume_data_source_validator_operation_count counter
+volume_data_source_validator_operation_count{result="test_result_1"} 1
+volume_data_source_validator_operation_count{result="test_result_2"} 2
+`
```

Comment:
So we don't have a metric about how long the operation takes, i.e., operation_total_seconds?

Reply:
The validator is just a webhook that runs are admission time. It has no idea what happens after that.

Information about how long population takes will be supplied by the populators themselves -- and support for that is going into the library. I started work on that last year then got side tracked, but I have to get it done for 1.24.